### DISCUSSION TITLE
Terrain layer error

### DISCUSSION THREAD

**Discussion by laijackylai** [visgl/deck.gl#5734] [Q&A] [answered]
I am trying to implement a simple terrain layer with the following code snippet

![code](https://user-images.githubusercontent.com/66239914/117233764-54ea6500-ae56-11eb-8b59-f3c0608194eb.png)

The url I passed into the elevationData is this image served by my local expressjs server

![hk_terrain_resized](https://user-images.githubusercontent.com/66239914/117234195-130dee80-ae57-11eb-8bf2-f6caa956e4a4.png)

However, I was getting this error:
`deck: Expected grid soze to be 2^n+1, got 2126`

I don't quite understand, because according to the [docs](https://deck.gl/docs/api-reference/geo-layers/terrain-layer#elevationdata), the url I passed to the elevationData prop should generate a simpleMeshLayer and shouldn't be getting this error?

Please help, thank you

**Comment by kylebarron** [comment]
The `TerrainLayer` passes the image to the `TerrainLoader` in loaders.gl. That currently uses [Martini](https://github.com/mapbox/martini) to create a mesh from an elevation heightmap. Martini only supports square images. The error you're seeing is created by Martini itself:
https://github.com/mapbox/martini/blob/9f9e2f58104b4d849d7509a2234cf3df3eff84e5/index.js#L6-L7

So with the current implementation this behavior is expected. It would be possible to switch the loaders.gl lib from Martini to [Delatin](https://github.com/mapbox/delatin), which doesn't have that error, but it's significantly slower.

**Comment by ibgreen** [comment, marked as answer]
@laijackylai Feel free to open an issue on loaders.gl. Can't promise we will get to it in the near future, but it could be interesting to offer an option to select which tesselator to use, and if so we could also automatically fall back if non-power-of-two.

@igorDykhta

**Reply by kylebarron** [reply]
I think it should be near-trivial to swap Delatin and Martini; it's just the bundle size impact that's the downside. Although apparently [Delatin is only 5.7kb minified](https://bundlephobia.com/result?p=delatin@0.2.0), so maybe including it and falling back to it for non-power-of-2 heightmaps is worth the bundle size impact.

**Reply by ibgreen** [reply]
> it's just the bundle size impact that's the downside

I have been working on making all loader modules dynamically loadable (which is already the case for worker loaders). This would dramatically reduce the ever-present bundle size concern, as loader code would only be loaded into the app the first time that specific loader is used.

This could potentially also allow us to offer all the loader objects in a `core` style module, as they don't take up much space, and not have to install and import each module in apps.

It is pretty close, but given my current workload I don't think it will make it into loaders.gl 3.0 unfortunately.

**Reply by laijackylai** [reply]
Thank you for the replies, I'll go ahead and open an issue on loaders.gl. Meanwhile, I think I'll give it a go at implementing Delatin into the terrain laoder myself.
